feat(Filters): validate inputs optionally with external func

youssefssc opened the pull request:
- Field level validation with optional custom `validate` functions and error messages
- Form level validations using `onError` prop for inputs and disabling of `Apply` button in case of errors

ajkl2533 commented on src/components/Filters/enums.ts:
```diff
@@ -0,0 +1,23 @@
+export const patterns = {
```
These patterns should be dropped from DS, we shouldn't export anything related to the actual usage of DS components. And the patterns should be implemented directly in the platform.

ajkl2533 replied:
At worst should be moved to mocks

PrincAm replied:
@youssefssc I'm not aware of use case to have predefined patterns. I would keep it as it was and define each pattern separately for particular story together with pattern message.

ajkl2533 commented on src/components/Filters/components/Input/Input.tsx:
```diff
@@ -44,25 +49,37 @@ export const StyledInput = styled.input`
+
+  const handleOnChange = (event) => {
     onChange(event);
+    if (isInvalid) {
```
Not sure it's a good idea to call validate function after the actual change. I'd say the onChange shouldn't happen when the field is invalid IMO

PrincAm replied:
`onChange` will change Filters state by wrong value and then Filters state is exposed by another public prop `onChange`, it shouldn't probably happen. I would move it into else after validation check. `Input` is controlled component so if `event.target` will be invalid then `onChange` will not be called and user will not be able to continue to input wrong values. I would say it is fine

youssefssc replied:
When I tried this, the field value could no longer be updated when invalid. It's because onChange is tied to `handleValueChange` in `FilterRow`. Any suggestions @PrincAm ?

PrincAm replied:
Not even by backspace? If not we have to leave it because we control the state as I wrote

PrincAm replied:
There could be also issue with copy pasting the value, we have to bear it in mind

ajkl2533 commented on src/components/Filters/components/Input/Input.tsx:
```diff
@@ -44,25 +49,37 @@ export const StyledInput = styled.input`
   maxLength,
-  pattern,
-  patternMessage,
+  pattern = patterns.string.pattern,
```
I would drop all default values related to patterns and let devs decide what they want and need.

PrincAm commented on src/components/Filters/components/Input/Input.tsx:
```diff
@@ -44,25 +49,37 @@ export const StyledInput = styled.input`
-  patternMessage,
+  pattern = patterns.string.pattern,
+  patternMessage = patterns.string.patternMessage,
+  validate = validateDefault,
```
[minor] Prop should be named `onValidate`. `validateDefault` then evaluate the pattern, right? I would name it `patternValidation` or something more descriptive

PrincAm commented on src/components/Filters/BottomBar/BottomBar.tsx:
```diff
@@ -22,6 +22,7 @@ const BottomBar: React.FC<BottomBarProps> = ({
   isLoading = false,
   hasUnappliedFilters,
   isCancelDisabled = false,
+  isApplyDisabled = false,
```
On wrong input value, it doesn't look `disabled` is changed to `true` 🤔

PrincAm commented on src/components/Filters/Filters.tsx:
```diff
@@ -96,6 +97,14 @@ const Filters: React.FC<FiltersProps> = ({
     }
   }, [filtersValues, fields]);
 
+  const validateFieldValues = () => {
```
Do we need it? `validationErrors` is always empty array

PrincAm commented on src/components/Filters/Filters.tsx:
```diff
@@ -96,6 +97,14 @@ const Filters: React.FC<FiltersProps> = ({
+    // TO DO: check validation for individual fields values
+    setHasInvalidValues(!!validationErrors.length);
+
+    return validationErrors;
```
Pls, what was the intention with returning of the array of errors? Can be removed?

PrincAm commented on src/components/Filters/Filters.tsx:
```diff
@@ -255,6 +262,7 @@ const Filters: React.FC<FiltersProps> = ({
       })}
       <BottomBar
         hasUnappliedFilters={hasUnappliedFilters}
+        isApplyDisabled={hasInvalidValues}
```
Be aware that Filters are using form validation with submit button inside `ButtonBar`. Then there is mismatch of all validations together, your custom and native. We want to get rid of the native validation message. Pls see the screenshot 
<img width="452" alt="Screen Shot 2021-04-21 at 11 21 52" src="https://user-images.githubusercontent.com/6930850/115530204-d9a08500-a293-11eb-8bf8-47bb801920a0.png">

PrincAm commented on src/components/Filters/Filters.types.ts:
```diff
@@ -24,6 +24,7 @@ interface ComponentProps {
   maxDate?: Date;
   units?: string;
   placeholder?: string | DateRangePlaceholderProps;
+  validate?: (target: HTMLInputElement) => string | boolean;
```
Pls, what is the use case for return of `string` by validation function? I would expect input value in, true or false out

PrincAm commented on src/components/Filters/components/Input/Input.tsx:
```diff
@@ -44,25 +49,37 @@ export const StyledInput = styled.input`
+  patternMessage = patterns.string.patternMessage,
+  validate = validateDefault,
   placeholder = 'String',
+  isDisabled = false,
```
Do we need to disable the input field?

PrincAm commented on src/components/Filters/helpers.ts:
```diff
@@ -0,0 +1,16 @@
+  const { target } = event;
+  const error = validate(target);
+  target.setCustomValidity(
+    target.validity.patternMismatch ? patternMessage : '',
```
We don't want to display native validation message. It should be removed

youssefssc replied:
If removed, the native validation message still displays on "Apply" but with default message "Please match the requested format". IMO it's better to have the same message as the error.

PrincAm commented on src/components/Filters/components/Input/Input.tsx:
```diff
@@ -44,25 +49,37 @@ export const StyledInput = styled.input`
+  const [isInvalid, setIsInvalid] = useState(false);
 
+  const handleOnValidate = (event) => {
+    const invalid = onValidate(event, validate, patternMessage);
```
It seems to me to be unnecessarily complicated. It will get the value, call validate, it will return good or bad and we will display error message and invalidate the apply button. Or I am wrong?

PrincAm commented on src/components/Filters/helpers.ts:
```diff
@@ -0,0 +1,16 @@
+  return !!target.validity.patternMismatch;
+};
+
+export const onValidate = (
```
Do we need to use this helper function?

youssefssc replied:
removed this func in commit before last

PrincAm reviewed (commented):
I left a few comment. Also please notice that styles of state button are broken.
<img width="968" alt="Screen Shot 2021-04-21 at 11 44 58" src="https://user-images.githubusercontent.com/6930850/115533470-0904c100-a297-11eb-99ce-da1c8d1bed23.png">

PrincAm commented:
@youssefssc Please notice, custom validation should be available for `TagsInput` and maybe other components as well. `Input` and `TagsInput` are needed, the rest we can cover later on.

youssefssc commented:
@PrincAm The rest of components' fields are strict in what the user can type or input (... Count, Integer, Number, Select, and dates ) I wasn't sure about Tags but will add the custom validations for it too.

youssefssc commented:
@PrincAm The validation for TagsInput are now in.

PrincAm commented on src/components/Filters/FilterRow/FilterRow.tsx:
```diff
@@ -154,9 +163,8 @@ const renderSelectComponent = (Component, value, onChange) => {
 };
 
-const renderComponent = (Component, value, onChange) => {
+const renderComponent = (Component, value, onChange, onError) => {
```
I think we don't need to pass `onError` to component without additional properties because all validations should be defined in component props object. Or I am missing some use case?

PrincAm commented on src/components/Filters/FilterRow/FilterRow.tsx:
```diff
@@ -166,10 +174,9 @@ const renderComponent = (Component, value, onChange) => {
   }
-
-  return <Component value={value} onChange={onChange} />;
+  return <Component value={value} onChange={onChange} onError={onError} />;
```
I think we don't need to pass onError to component without additional properties because all validations should be defined in component props object. Or I am missing some use case?

PrincAm commented on src/components/Filters/Filters.tsx:
```diff
@@ -58,7 +58,7 @@ const getDefaultState = ([firstField]: Field[]) => {
 };
 
-const Form = styled.form`
+const Form = styled.div`
```
Could we rename `Form` to something else if it is not from anymore? Also you can `FlexContainer` instead

PrincAm commented on src/components/Filters/components/Count/Count.tsx:
```diff
@@ -9,23 +11,34 @@ const Count: React.FC<NumberProps> = ({
 
   return (
-    <StyledNumber
-      max={max}
```
We can't use `min`, `max`, `step` native attributes?

youssefssc replied:
They would be useless to add here. The range validation is done in `validateNumber` so we can show our custom errors.

PrincAm replied:
I understand how it works. I thought we agreed we will use native validation and just display custom message for sake of simplicity, so it is not possible?

youssefssc replied:
It is possible but does nothing as I understand

PrincAm commented on src/components/Filters/components/Count/Count.tsx:
```diff
@@ -9,23 +11,34 @@ const Count: React.FC<NumberProps> = ({
   };
+  const handleOnChange = (event) => {
+    onChange(event);
+    const error = !validateNumber(event.target.value, min, max);
```
If `error` has boolean type, use `is` or `has` prefix in name

PrincAm commented on src/components/Filters/components/Input/Input.tsx:
```diff
@@ -52,17 +63,23 @@ const Input: React.FC<InputProps> = ({
+
+  const handleOnValidate = (event) => {
+    const { target } = event;
+    const error = onValidate(target);
```
Same here, `onValidate` should return boolean and name should have boolean prefix

youssefssc replied:
I updated the names for all inputs but in the Input and TagsInput case we need the whole target because 
```
validatePattern = (target: HTMLInputElement): boolean => {
  return !!target.validity.patternMismatch;
};```

PrincAm replied:
Okey. `onValidate` is not custom validation here, it mislead me a bit.

PrincAm commented on src/components/Filters/components/Count/Count.tsx:
```diff
@@ -9,23 +11,34 @@ const Count: React.FC<NumberProps> = ({
   min,
   max,
   placeholder = 'Count',
+  patternMessage = 'Use only integers',
```
I would rename all `patternMesage`s to something more general like `errorMessage` since it is not bound only to pattern. Maybe will be needed update of FE if we will change the API of filters

PrincAm commented on src/components/Filters/helpers.ts:
```diff
@@ -0,0 +1,16 @@
+export const validatePattern = (target: HTMLInputElement): boolean => {
+  return !!target.validity.patternMismatch;
```
[minor] You can omit `return`

PrincAm commented on src/components/Filters/helpers.ts:
```diff
@@ -0,0 +1,16 @@
+  return !!target.validity.patternMismatch;
+};
+
+export const validateNumber = (
```
It should be covered by unit test

PrincAm commented:
@youssefssc Please notice I can apply filters even if there are some fields invalid
<img width="1230" alt="Screen Shot 2021-04-26 at 14 24 33" src="https://user-images.githubusercontent.com/6930850/116082011-35f01400-a69b-11eb-8787-12821224b84c.png">

PrincAm commented:
@youssefssc One more observation. I would mark empty value as valid to make it consistent because it is same like on the start. Furthermore, if you click on Apply, all filters with empty values are removed.
<img width="1232" alt="Screen Shot 2021-04-26 at 14 31 28" src="https://user-images.githubusercontent.com/6930850/116083247-b4998100-a69c-11eb-9c34-e2210844549c.png">

youssefssc commented:
@PrincAm I added fixes for the last two bugs in screenshots. So now empty values are valid and we have an array `validsValues` which must all be `true` to enable the Apply button.  It's not complete yet as I am not sure how and where exactly to update the size of this array when the number of filters change but please take a look and let me know if this is going in the right direction. Thanks.

PrincAm commented:
@youssefssc We have two places where we keep information about inputs validity. One in each component `isInvalid` and second in Filters `validsValues` (`currentFieldIndex`). We should avoid that and have only one source of truth. I would remove local states from components, kept them simple as they were and move state with errors to `Filters` component similarly how filters values are managed. Filters errors can be object where key will be an index of the row and value boolean, e.g. `{ 0: true, 2: true }`

PrincAm commented:
@youssefssc Also pls notice there are some prop types warnings when used validation
<img width="934" alt="Screen Shot 2021-04-28 at 12 58 59" src="https://user-images.githubusercontent.com/6930850/116394864-b5105400-a823-11eb-89eb-4ac9889816fc.png">

PrincAm commented on src/components/Filters/Filters.tsx:
```diff
@@ -97,6 +93,16 @@ const Filters: React.FC<FiltersProps> = ({
   }, [filtersValues, fields]);
 
+  const handleError = (hasError, index) => {
+    // TO DO: handle all validations in FilterRow in handleValueChange
```
I think this TODO can be removed

PrincAm commented on src/components/Filters/Filters.tsx:
```diff
@@ -97,6 +93,16 @@ const Filters: React.FC<FiltersProps> = ({
 
+  const handleError = (hasError, index) => {
+    // TO DO: handle all validations in FilterRow in handleValueChange
+    const newValidsValues = validsValues;
```
It should be a copy

PrincAm commented on src/components/Filters/Filters.tsx:
```diff
@@ -78,6 +72,7 @@ const Filters: React.FC<FiltersProps> = ({
   const [filtersValues, setFiltersValues] = useState<Array<Filter>>(null);
   const [isDefaultState, setIsDefaultState] = useState(true);
   const [hasUnappliedFilters, setHasUnappliedFilters] = useState(false);
+  const [validsValues, setValidsValues] = useState<boolean[]>([true]);
```
Isn't `validValues` correct name?

PrincAm replied:
Shouldn't be the default value an empty array? Or it should be derived from `state`

youssefssc replied:
the default is an array of one true element because it's one Field / FilterRow. We can move it to the first useEffect but it's the same

PrincAm replied:
If `state` passed down to Filters will include 5 "filters" definitions then it should include five `true`s. Therefore it should be derived from state.length

PrincAm commented:
If I remove row with invalid value then following row is marked as invalid. It's not correct
<img width="1126" alt="Screen Shot 2021-04-29 at 15 20 49" src="https://user-images.githubusercontent.com/6930850/116557501-c7ab8b80-a8fe-11eb-937d-8786a3311b22.png">

PrincAm commented on src/components/Filters/components/TagsInput/TagsInput.tsx:
```diff
@@ -127,56 +131,63 @@ const TagsInput: React.FC<TagsInputProps> = ({
+    } = event;
 
-    if (value && !validity.patternMismatch) {
+    if (value && !isInvalid) {
```
I'm able to add invalid tag value on blur :(

PrincAm replied:
But Apply button is disabled so we can fix that later, imo

PrincAm replied:
Story "Tags Input Validation", insert wrong value and click outside the input. Tag with incorrect value is added. You can compare this behavior with adding of wrong value on enter.

youssefssc replied:
I don't have this behavior on blur  :S

PrincAm replied:
It is fixed

youssefssc commented:
I addressed the new comments and implemented the update for validValues on filter add/remove/clear all
Also removed isDisabled for Input 

@PrincAm Let me know if you think this PR is ready to be merged, and if there is a small bug let me know and I will create a ticket for it

PrincAm commented:
I can't insert correct value when used `min`, `max` validation
<img width="793" alt="Screen Shot 2021-04-29 at 16 56 05" src="https://user-images.githubusercontent.com/6930850/116572056-ea906c80-a90b-11eb-8336-7f5956ddb101.png">

PrincAm commented on src/components/Filters/components/Input/Input.tsx:
```diff
@@ -32,9 +35,9 @@ export const StyledInput = styled.input`
-  &:invalid {
-    ${stateStyles}
-    border-color: ${getFormStyle('invalidBorderColor')};
+  &:disabled {
```
Do we need this styles for disabled state? I don't think we are disabling the input

PrincAm reviewed (approved):
LGTM
